ports/libsimpleservo/capi/src/lib.rs

```diff
@@ -16,12 +16,118 @@ use std::ffi::{CStr, CString};
+    let mut _secattr: minwinbase::SECURITY_ATTRIBUTES = unsafe { mem::zeroed() };
+
+    _secattr.nLength = mem::size_of::<minwinbase::SECURITY_ATTRIBUTES>() as u32;
+    _secattr.bInheritHandle = shared::minwindef::TRUE;
+    _secattr.lpSecurityDescriptor = shared::ntdef::NULL;
+
+    unsafe {
+        if namedpipeapi::CreatePipe(&mut _h_read_pipe, &mut _h_write_pipe, &mut _secattr, 512) == 0
+        {
+            return false;
+        }
+
+        if processenv::SetStdHandle(winbase::STD_OUTPUT_HANDLE, _h_write_pipe) == 0 ||
+            processenv::SetStdHandle(winbase::STD_ERROR_HANDLE, _h_write_pipe) == 0
+        {
+            return false;
+        }
+
+        if handleapi::SetHandleInformation(
+            _h_read_pipe,
+            winbase::HANDLE_FLAG_INHERIT,
+            winbase::HANDLE_FLAG_INHERIT,
+        ) == 0 ||
+            handleapi::SetHandleInformation(
+                _h_write_pipe,
+                winbase::HANDLE_FLAG_INHERIT,
+                winbase::HANDLE_FLAG_INHERIT,
+            ) == 0
+        {
+            return false;
+        }
+
+        let _h_read_pipe_fd = libc::open_osfhandle(_h_read_pipe as libc::intptr_t, libc::O_RDONLY);
+        let _h_write_pipe_fd =
+            libc::open_osfhandle(_h_write_pipe as libc::intptr_t, libc::O_WRONLY);
+
+        if _h_read_pipe_fd == -1 && _h_write_pipe_fd == -1 {
+            return false;
+        }
+
+        // 0 indicates success.
+        if libc::dup2(_h_write_pipe_fd, 1) != 0 && libc::dup2(_h_write_pipe_fd, 2) != 0 {
+            return false;
+        }
+
+        // If SetStdHandle(winbase::STD_OUTPUT_HANDLE, hWritePipe) is not called prior,
+        // this will fail.  GetStdHandle() is used to make certain "servo" has the stdout
+        // file descriptor associated.
+        let _h_stdout = processenv::GetStdHandle(winbase::STD_OUTPUT_HANDLE);
+        if _h_stdout == handleapi::INVALID_HANDLE_VALUE || _h_stdout == shared::ntdef::NULL {
+            return false;
+        }
+
+        // If SetStdHandle(winbase::STD_ERROR_HANDLE, hWritePipe) is not called prior,
+        // this will fail.  GetStdHandle() is used to make certain "servo" has the stderr
+        // file descriptor associated.
+        let _h_stderr = processenv::GetStdHandle(winbase::STD_ERROR_HANDLE);
+        if _h_stderr == handleapi::INVALID_HANDLE_VALUE || _h_stderr == shared::ntdef::NULL {
+            return false;
+        }
+
+        // Spawn a thread.  The thread will redirect all STDOUT and STDERR messages
+        // to OutputDebugString()
+        let _handler = thread::spawn(move || {
+            loop {
+                const BUF_LENGTH: usize = 1024;
+                let mut read_buf: [i8; BUF_LENGTH] = mem::zeroed();
+
+                let _result = {
+                    libc::read(
+                        _h_read_pipe_fd,
+                        read_buf.as_mut_ptr() as *mut _,
+                        read_buf.len() as u32,
```

Comment:
We should pass `read_buf.len() as u32 - 1` to ensure that there's always a terminating null byte, since OutputDebugString requires that.